utils/repostate/statemanager.go

```diff
@@ -0,0 +1,194 @@
+// The state is constructed as a linked prefix tree, where every node has pointers to its children and parent.
+// While traversing over the repository, contents found are added to their respecting node, and those handled are then removed.
+// Each node has one of three states:
+//   1. Unexplored / Partially explored - NOT all contents of the directory were found and added to its struct (Marked by !DoneExploring).
+//   2. Fully explored - All contents of the directory were found, but NOT all of them handled (Marked by DoneExploring && !Completed).
+//   3. Completed - All contents found and handled (Marked by Completed).
+// In the event of a node reaching completion, a tree collapsing may occur in order to save space:
+// The node will mark itself completed, and will then notify the parent to check completion as well.
+type RepoStateManager struct {
+	Root    *Node `json:"root,omitempty"`
+	repoKey string
+	// File path to save and load state to/from.
+	stateFilePath string
+}
+
+// Used as an empty value for the FilesNames map, where we only use keys.
+var emptyValue struct{}
+
+type Node struct {
+	Name          string              `json:"name,omitempty"`
+	Parent        *Node               `json:"-"`
+	Children      map[string]*Node    `json:"children,omitempty"`
+	FilesNames    map[string]struct{} `json:"files,omitempty"`
+	Completed     bool                `json:"completed,omitempty"`
+	DoneExploring bool                `json:"done_exploring,omitempty"`
+}
+
+// Tries to load a repo state from the provided repoStatePath.
+// If successful, returns the state and created=false.
+// Otherwise, generates a new repo state and returns it with created=true.
+func LoadOrCreateRepoStateManager(repoKey, repoStatePath string) (sm *RepoStateManager, created bool, err error) {
+	exists, err := fileutils.IsFileExists(repoStatePath, false)
+	if err != nil {
+		return nil, false, err
+	}
+
+	stateManager := new(RepoStateManager)
+	if exists {
+		content, err := fileutils.ReadFile(repoStatePath)
+		if err != nil {
+			return nil, false, err
+		}
+
+		err = json.Unmarshal(content, stateManager)
+		if err != nil {
+			return nil, false, errorutils.CheckError(err)
+		}
+		// Parent references cannot be saved to file due to cyclic references. We add them manually here.
+		stateManager.Root.addParentsReferences(nil)
+	} else {
+		stateManager.Root = createNewNode(".", nil)
+		created = true
+	}
+
+	stateManager.repoKey = repoKey
+	stateManager.stateFilePath = repoStatePath
+	return stateManager, created, nil
+}
+
+func (sm *RepoStateManager) SaveToFile() error {
+	content, err := json.Marshal(sm)
+	if err != nil {
+		return errorutils.CheckError(err)
+	}
+	return errorutils.CheckError(ioutil.WriteFile(sm.stateFilePath, content, 0644))
+}
+
+// Returns the node corresponding to the directory in the provided relative path. Path should be provided without the repository name.
+func (sm *RepoStateManager) LookUpNode(relativePath string) (*Node, error) {
```

Comment:
Don't we have an issue of thread safety here? If multiple go routine can call this API simultaneously, we should find the best and most efficient way to protect the critical sections of the data structure.
If indeed we have an issue here, and we need to modify the code, I'd be happy to revierw the code again after those changes.